### ISSUE TITLE
Portable Version Improvements

### ISSUE THREAD

**Issue by 9z0** [ransome1/sleek#131]
1. Provide it as an extractable archive (zip/7z) because the current one has to be extracted on the fly before use.

2. Save (at least the most relevant setting files) in the application directory. You can check other electron projects like Visual Studio Code.

**Comment by ransome1** [comment]
@9z0 as for 1. What exactly do you mean with `extracted on the fly`?

As for 2. please check if https://github.com/ransome1/sleek/issues/72#issuecomment-821065649 is any help. I'm not familiar with Windows, but as far as I understood back then, a portable file is designed to create it's own temp folder everytime it is replaced with a new one.

To save sleeks settings in a permanent location the setup/installer file can be used, which is part of the release. I assume this is what Visual Studio does too, as the download is not a portable but an installer, if I am not mistaking: https://code.visualstudio.com/docs/?dv=win32user

**Comment by 9z0** [comment]
> What exactly do you mean with `extracted on the fly`?

I meant that **every** time you double click the "portable" (no install) version, it extracts the contents into a temp folder. This slows launch times considerably.

Portability (in that sense) in Windows isn't just about avoiding needing admin privileges, but mainly:
1. Running from a moving location (like a USB drive).
2. Saving settings in a place that is always accessible from that moving location. Practically that means a relative path under the application directory (like `./Config`)
3. A third point (closely-related although not essential to that understanding of portability) is leaving a minimum of traces on the host PC. This means avoiding temporary folders and files if possible and removing them afterwards if not possible.

You can summarize the philosophy by "everything within the relative directory."


> please check if [#72 (comment)](https://github.com/ransome1/sleek/issues/72#issuecomment-821065649) is any help.

I have skimmed the issue before, and I read from your linked post down.

Electron is a bit notorious for not being portable-friendly, but it has been done in multiple Electron-dependent projects. For example, Visual Studio **Code** has a [portable mode](https://code.visualstudio.com/docs/editor/portable). So do [Atom](https://github.com/atom/atom), [TagSpaces](https://github.com/tagspaces/tagspaces), [Mark Text](https://github.com/marktext/marktext), etc.

Perhaps they can give some technical insights on the matter. :)

Thanks

**Comment by ransome1** [comment]
@9z0, if you say `Practically that means a relative path under the application directory (like ./Config)` this would mean that the config file is stored in a sub directory relative to where the portable is located? Like for instance
App: `c:/home/user/applications/sleek.exe`
Config: `c:/home/user/applications/config/sleek.json`

**Comment by 9z0** [comment]
Yes.

Some solutions use a directory like the example you gave. Others place the setting file(s) right next to the executable:
`c:/home/user/applications/sleek.exe`
`c:/home/user/applications/sleek.json`

I feel that the first is cleaner, especially with multiple config files.

There are also multiple ways to "detect" that the user wants to use the portable version:
- Use a marker file. For example, an empty file named `portable` next to the executable.
- Test for the settings file/folder in the application directory. If exists, use it and don't use the user profile.
- Launching the app with an argument. `sleek.exe --portable`.
- Release a special version of the executable that always has a portable behavior.
- Some tools assume that any install outside the default directory (%PROGRAMFILES%) is a portable install.
- Some tools depend on the name of the executable! Like if you rename `sleek.exe` to `sleek-por.exe`, it will detect that and behave portably!

I prefer the first solution, because:
1. It allows the developer to provide the portable version in an extractable archive where the marker exists by default.
2. If you included the settings file (solution 2) in the archive to indicate portability, the user would have to be careful not to overwrite their own modified settings file from a previous "install"/extraction. The empty `portable` file can be safely overwritten.
3. Solution 3 requires an extra step/shortcut file/etc. You'd have to always remember not to launch `sleek.exe` directly.
4. Solution 4 is more work for the developer and (along with the last two solutions) feels against good, sane practices of preferring the easier, modular solution.

**Comment by ransome1** [comment]
I'll see what can be done about the config file issue.

About the portable topic I want to be honest; I have little to no experience with Windows, portable files and so on and am not sure what can be done about it. The portable is being built with Electron Builder and the configuration (https://www.electron.build/configuration/win) is quite limited here. So I am not sure if the solution you're proposing (and which I don't understand ;)) can be implemented.

**Comment by 9z0** [comment]
There are also subtle behavior changes/options with portable use to make the experience better.

Supporting relative paths not just for settings files but also working files. You may want to have your todo.txt files(s) in the same USB drive. So, if you keep them under the sleek folder, sleek can automatically assign these files relative paths in its settings.
For example, instead of:
` G:\My Apps\sleek\tasks\todo.txt`
it would be saved in the preferences as
`.\tasks\todo.txt` or `tasks\todo.txt`
but
`G:\todos\todo.txt` (in case you want them more accessible to you or if you have other todo.txt apps) would be either:
  - saved as an absolute path
  - offered to the user to be saved as a path relative to sleek **or** the drive:
    - `..\..\todos\todo.txt`
    - `{root_letter}\todos\todo.txt`

This is just an example of some of the purposes of portable use and how it is approached. Don't take them as an urgent request, but they are nice to have. :)

**Comment by 9z0** [comment]
> About the portable topic I want to be honest; I have little to no experience with Windows, portable files and so on and am not sure what can be done about it. The portable is being built with Electron Builder and the configuration (https://www.electron.build/configuration/win) is quite limited here. So I am not sure if the solution you're proposing (and which I don't understand ;)) can be implemented.

No worries, the relative path use of the important config file(s) is 80% of the matter. The rest was just to shed some light on the matter. :)

The simplest solution is to have sleek check for the existence of a folder called `config` (for example) on startup. If it exists, save settings there.

The linked Electron projects above have that already implemented which that can be applied here. I hope it is reasonable enough to implement. :)

Thanks

**Comment by ransome1** [comment]
@9z0, I don't think just replacing it would be a good idea, as quite some users want to use the portable .exe build and don't mind a longer startup time. But I added a zipped version and persisted the location of the user data.

For the **portable** build the setup would be
`c:/pathToPortable/sleek-1.0.5-3.exe`
`c:/pathToPortable/config/sleek/user-preferences.json`

For the **zipped** version
`c:/pathToUnzippedVersion/sleek.exe`
`c:/pathToUnzippedVersion/user-preferences.json`

Can you please check this pre-release, if it is working as expected on a real Windows production environment?https://github.com/ransome1/sleek/releases/tag/v1.0.5-3

I got two questions though:
1. About the zipped version: How is the common behavior of Windows users when they download an update of a zipped version? Do they just extract the content of the new zip file into the old folder an replace the old files with the new ones? So can I assume the user-preferences.json won't be overwritten in this process?
2. When users make use of a portable .exe, do they actually expect, that the user data is being created into the same directory, where the portable .exe is located? Or do they expect user data to be placed in a central and Windows specific location for user data?

**Comment by 9z0** [comment]
> @9z0, I don't think just replacing it would be a good idea, as quite some users want to use the portable .exe build and don't mind a longer startup time. But I added a zipped version and persisted the location of the user data.
> 
> For the **portable** build the setup would be
> `c:/pathToPortable/sleek-1.0.5-3.exe`
> `c:/pathToPortable/config/sleek/user-preferences.json`
> 
> For the **zipped** version
> `c:/pathToUnzippedVersion/sleek.exe`
> `c:/pathToUnzippedVersion/user-preferences.json`

This is fine by me, although `pathToPortable/config/sleek/user-preferences.json` is a bit of an atypical location for portable use. For me it implies a common `config` folder like `~/.config` in Linux under which different apps save their settings one folder each (`config/{app}`). Even if the app is just one executable file, the customary behavior (for a protable app) is to save settings under `config` directly. This is a fairly trivial point though.

> Can you please check this pre-release, if it is working as expected on a real Windows production environment?https://github.com/ransome1/sleek/releases/tag/v1.0.5-3

I copied over my current `user-preferences.json` and it seems to work fine. Other Electron settings are still saved under the user profile, but at least the important settings are portable now 😃.

>1. About the zipped version: How is the common behavior of Windows users when they download an update of a zipped version? Do they just extract the content of the new zip file into the old folder an replace the old files with the new ones? So can I assume the user-preferences.json won't be overwritten in this process?

Pretty much. Extract over the old. You seem to have hard-coded the portability of sleek into the executable, so there is no need for a "portability marker" or packaging an empty/default `user-preferences.json` inside the archive. If you unify the executable compilation at some point in the future, I suggest you use the `portable` file marker as mentioned before since it is the safest.

>2. When users make use of a portable .exe, do they actually expect, that the user data is being created into the same directory, where the portable .exe is located? Or do they expect user data to be placed in a central and Windows specific location for user data?

A user of portable software **does** expect (this is the whole point) it to save settings somewhere **within** the application folder. Either:
- directly next to the main executable:
```
sleek/sleek.exe
sleek/config.json
sleek/favorites.json
sleek/bookmarks.json
```
- inside a folder which is directly next to the main executable:
```
sleek/sleek.exe
sleek/config/config.json
sleek/config/favorites.json
sleek/config/bookmarks.json
```

The second is better with multiple files (to facilitate backup for example).

The default (non-portable) behavior in Windows is to save settings per user profile:
- Under %USERPROFILE%: for example, sleek (setup version) saves under C:\Users\\{username}\AppData\Roaming\sleek
- In the registry.
- In one of multiple other locations under C:\Users\\{username}

---

Most users don't actually care as far as I can tell. They just go with the defaults.

The main benefit of the default %USERPROFILE% behavior is to allow multiple users of the same system to install an application once, but have a set of settings each.

The main benefit of the portable behavior is to have the settings always accompany the app independent of the computer you run it on. If you copy the application folder, you always take the settings with it.

---

I **encourage** you to read (like, actually give it a read) these quick definition/FAQ points about portable software on Windows:

[Definition](https://www.portablefreeware.com/about.php)
[FAQ](https://www.portablefreeware.com/faq.php) (the first five points.)

**Comment by ransome1** [comment]
I'm not sure if we are talking about the same thing ;) When I refer to the portable .exe, I mean the exe file, that electron-builder is creating when it is advised to build a "portable". This one here to be precise: https://github.com/ransome1/sleek/releases/download/v1.0.5-3/sleek-1.0.5-3.exe

It's an archive that among others contains the actual executable. In this case I still don't understand where a Windows user is expecting the config file to be located, if you say it's atypical in the way I configured it now. Of course the config file could be placed within this archive, but would be overwritten everytime a user is downloading a new version of sleek.

**Comment by 9z0** [comment]
I see.

I don't know!

- There aren't many examples I know of for this scenario.
- Since Electron has problems with portability, I doubt their interpretation and use of the term is the same [as the community](https://en.wikipedia.org/wiki/Portable_application).
- I would guess that Electron uses "portable" to mean "no-setup" or "no-install" rather than saving settings portably.
- The one actual application I can think of that behaves like that and is actually portable is [Joplin](https://github.com/laurent22/joplin/releases). If you run "JoplinPortable.exe":
    - it extracts the contents to a temp folder.
    - it saves settings in a folder `JoplinProfile` next to JoplinPortable.exe itself, not in the temp folder.
- Since users who are aware of, and prefer, portable applications are more tech-savvy than the average user, they tend to be sensitive to something like extract-on-every-run. Actually, many of them hate Electron apps (with a passion!); since they view them as bloated and resource-hungry and are often not portable.


So, in short, sleek-1.0.5-3.exe seems to fall in a weird place between:
- users who don't care about portability but prefer/need a no-install version.
- those who care about portability but can't find a better option (which is now provided by the zip file).

The main value provided by sleek-1.0.5-3.exe (no need for installation/admin privileges) is also provided by the zipped version. You can keep it for existing users, but I doubt many would miss it if it was removed.

**Comment by ransome1** [comment]
Ok, I see, slowly I seem to get this whole portable thing. By the way, a couple of weeks ago another sleek user mentioned https://portableapps.com/ as an interesting way of distributiing sleek. I don't have enough time to maintain another channel, but if maintaining sleek on portableapps.com would be interesting for you, I would be happy to support you if or where ever I could.

Anyhow, if this issue is solved, please do me a favor and close it, so we keep the issue list clean and tidy ;)

**Comment by 9z0** [comment]
I use portableapps software, but I never delved into the inner workings of it. :(

There is the whole field of "portablizers" which intercept the settings-related calls/operations of a non-portable app and direct those to that application's folder.

I don't think any of those would be needed if you can carry over the rest of Electron settings into sleek's folder like the examples mentioned [above](#issuecomment-850371063). If you can do that, you don't have to maintain anything.

Until then, I do consider the bare minimum of portability met and can close the issue.

👍
